Added text selection support to TextFieldWidget.

joealam opened the pull request:
Implements feature requested in issue #2794.
Use Shift and navigation key (cursor, home, end) to select a portion of
text, and replace/delete/cut as appropriate.
Also provides support for selection with mouse (click and drag)

Proposed changelog entry:
Added text selection and editing support in text boxes.

pchote commented:
This looks really great, thanks for the PR!

At a first look I have a couple of quick comments, and will hopefully follow up in more detail within the next few days:
* OpenRA code style omits braces around single line blocks
* Copying isn't working on macOS
* We may want to tweak the selection color to fit the color palette in each mod

CatGirls420-NerevarII commented:
Not all heroes wear capes!

pchote reviewed (commented):
Code changes and functionality LGTM aside from my points above.

MunWolf commented on OpenRA.Mods.Common/Widgets/TextFieldWidget.cs:
```diff
@@ -244,18 +298,47 @@ public override bool HandleKeyPress(KeyInput e)
+						&& !string.IsNullOrEmpty(Text))
+					{
+						// If a portion of text is selected, copy it to the clipboard
+						if (selectionStartIndex != -1)
```
Why is this not a part of the if statement above?

MunWolf commented:
Other then that one question the code looks good from my end.

joealam commented:
It's just layed out the same way as the Cut case for easy readability, but I'm happy to change it if preferred?

pchote commented:
LGTM, just a couple more small requests:

* Can you please merge the two fixup commits into the first one?
* Keeping the old behaviour where the copy/cut commands will act on the whole text box if nothing is selected feels wrong now that we have proper selection support.  Can we remove that?
* I suggest the following mod-specific highlight colors to better fit with the UI (merge into the first commit):
   * cnc: `800000`
   * ra: `562020`
   * ts: `1a1a1a`
   * d2k: `7f4d29`

joealam committed:
Added text selection and copy support to TextFieldWidget.

Use Shift and navigation key (cursor, home, end) to select a portion of
text, and replace/delete/cut as appropriate.
Also provides support for selection with mouse (click and drag)

pchote commented:
I hope you don't mind @joealam, but I have made the changes above and pushed over your branch so we can move ahead with this.

pchote reviewed (approved):
LGTM and works as advertised.

Smittytron reviewed (approved):
Works for me

joealam commented:
@pchote I don't mind at all! It's ideal in fact, is I'm going to be away from my PC for a while so wouldn't have had chance to do them myself.

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7c9cdfed4e8fd3c5b5bac4948f517ad016c2125d...5ab52260de2614784a104f1bca289fcbbed7d815)

Mailaender commented:
You can collect your bounty at https://www.bountysource.com/issues/136949-word-jump-and-selection-with-ctrl-shift-arrow-key-in-lobby now.
